postgresql/postgresql_plugin.cc

```diff
@@ -0,0 +1,832 @@
+
+    }
+
+    /** Return a row as an expression value.  Default forwards to the matrix
+    view's getRow() function.
+    */
+    virtual ExpressionValue getRowExpr(const RowName & row) const override
+    {
+        auto conn = startConnection();
+        string selectString = "SELECT * FROM ";
+        selectString += config_.tableName +
+                        " WHERE " + 
+                        config_.primaryKey + 
+                        " = '" + 
+                        row.toUtf8String().rawString()
+                        + "'";
+
+        POSTGRESQL_VERBOSE(cerr << "postgress row select: " << endl;)
+        POSTGRESQL_VERBOSE(cerr << selectString << endl;)
+
+        auto res = PQexec(conn, selectString.c_str());
+        if (PQresultStatus(res) != PGRES_TUPLES_OK) {
+            string errorMsg(PQresultErrorMessage(res));
+            PQclear(res);
+            PQfinish(conn);
+            throw HttpReturnException(400, "Could not select from postgreSQL: ", errorMsg);
+        }
+
+        POSTGRESQL_VERBOSE(cerr << "row fetched from " << config_.tableName << " successfully!" << endl;)
+
+        int nfields = PQnfields(res);
+        int ntuples = PQntuples(res);
+
+        POSTGRESQL_VERBOSE(cerr << ntuples << "," << nfields << endl;)
+
+        std::vector<std::tuple<ColumnName, CellValue, Date> > rowValues;
+        if (ntuples > 0) {
+            std::vector<std::tuple<ColumnName, CellValue, Date> > cols;
+            for(int j = 0; j < nfields; j++) {
+                rowValues.emplace_back(ColumnName(PQfname(res, j)), getCellValueFromPostgres(res, 0, j), Date::Date::notADate());
+                POSTGRESQL_VERBOSE(printf("[%d,%d] %s %s\n", 0, j, PQgetvalue(res, 0, j), PQfname(res, j));)
+            }       
+        }
+
+        PQclear(res);
+        PQfinish(conn);
+
+        return ExpressionValue(rowValues);
+
+    }
+
+    /** Return wheter or not all columns names and info are known.
+    */
+    virtual bool hasColumnNames() const { return false; }
+
+};
+
+/*****************************************************************************/
+/* POSTGRESQL RECORDER DATASET                                                */
+/*****************************************************************************/
+
+struct PostgresqlRecorderDatasetConfig {
+    string databaseName;
+    int port;
+    string host;
+    string tableName;
+    bool createTable;
+
+    PostgresqlRecorderDatasetConfig() {
+        databaseName = "database";
+        port = 1234;
+        host = "localhost";
+        tableName = "mytable";
+        createTable = false;
+    }
+};
+
+DECLARE_STRUCTURE_DESCRIPTION(PostgresqlRecorderDatasetConfig);
+DEFINE_STRUCTURE_DESCRIPTION(PostgresqlRecorderDatasetConfig);
+
+PostgresqlRecorderDatasetConfigDescription::
+PostgresqlRecorderDatasetConfigDescription()
+{
+    addField("databaseName", &PostgresqlRecorderDatasetConfig::databaseName, "Name of the database to connect to.");
```

Comment:
We can use a base config with the PostgreSQL _coordinates_ and derive from it for the other configs.  This would avoid the string duplications and make maintenance easier.